src/Stack/SourceMap.hs

```diff
@@ -0,0 +1,245 @@
+    => ActualCompiler
+    -> RIO env (Map PackageName GlobalPackage)
+globalsFromDump compiler = do
+    let pkgConduit =
+            conduitDumpPackage .|
+            CL.foldMap (\dp -> Map.singleton (dpGhcPkgId dp) dp)
+        toGlobals ds = Map.fromList $ map toGlobal $ Map.elems ds
+        toGlobal d =
+            ( pkgName $ dpPackageIdent d
+            , GlobalPackage (pkgVersion $ dpPackageIdent d))
+    toGlobals <$> ghcPkgDump (whichCompiler compiler) [] pkgConduit
+
+globalsFromHints ::
+       HasConfig env
+    => ActualCompiler
+    -> RIO env (Map PackageName GlobalPackage)
+globalsFromHints compiler = do
+    ghfp <- globalHintsFile
+    mglobalHints <- loadGlobalHints ghfp compiler
+    case mglobalHints of
+        Just hints -> pure $ Map.map GlobalPackage hints
+        Nothing -> do
+            logWarn $ "Unable to load global hints for " <> RIO.display compiler
+            pure mempty
+
+toActual ::
+       (HasConfig env)
+    => SMWanted
+    -> WithDownloadCompiler
+    -> ActualCompiler
+    -> RIO env SMActual
+toActual smw downloadCompiler compiler = do
+    allGlobals <-
+        case downloadCompiler of
+            WithDownloadCompiler -> globalsFromDump compiler
+            SkipDownloadCompiler -> globalsFromHints compiler
+    let globals =
+            allGlobals `Map.difference` smwProject smw `Map.difference` smwDeps smw
+    return
+        SMActual
+        { smaCompiler = compiler
+        , smaProject = smwProject smw
+        , smaDeps = smwDeps smw
+        , smaGlobal = globals
+        }
+
+checkFlagsUsedThrowing ::
+       (MonadIO m, MonadThrow m)
+    => Map PackageName (Map FlagName Bool)
+    -> FlagSource
+    -> Map PackageName ProjectPackage
+    -> Map PackageName DepPackage
+    -> m ()
+checkFlagsUsedThrowing packageFlags source prjPackages deps = do
+    unusedFlags <-
+        forMaybeM (Map.toList packageFlags) $ \(pname, flags) ->
+            checkFlagUsed (pname, flags) source prjPackages deps
+    unless (null unusedFlags) $
+        throwM $ InvalidFlagSpecification $ Set.fromList unusedFlags
+
+checkFlagUsed ::
+       MonadIO m
+    => (PackageName, Map FlagName Bool)
+    -> FlagSource
+    -> Map PackageName ProjectPackage
+    -> Map PackageName DepPackage
+    -> m (Maybe UnusedFlags)
+checkFlagUsed (name, userFlags) source prj deps =
+    let maybeCommon =
+          fmap ppCommon (Map.lookup name prj) <|>
+          fmap dpCommon (Map.lookup name deps)
+    in case maybeCommon  of
+        -- Package is not available as project or dependency
+        Nothing ->
+            pure $ Just $ UFNoPackage source name
+        -- Package exists, let's check if the flags are defined
+        Just common -> do
+            gpd <- liftIO $ cpGPD common
+            let pname = pkgName $ PD.package $ PD.packageDescription gpd
+                pkgFlags = Set.fromList $ map PD.flagName $ PD.genPackageFlags gpd
+                unused = Set.difference (Map.keysSet userFlags)
```

Comment:
There is a handy function `Map.withoutKeys` that could be used here instead, which would avoid unnecessary conversion to `Set` until necessary